### PR TITLE
Python 2 and 3 compatibility

### PR DESCRIPTION
**Pull request by FredrikAppelros**
Fix the broken Python 3 support, and make sure that the code remains compatible with Python 2.

### CODE REVIEW

**Comment by FredrikAppelros** [comment]
This fixes issue #85 while remaining compatible with Python 2. It has been tested using Python 2.7.10 and Python 3.4.3.

But please, do take a close look before merging, as I have only tried to make the minimal amount of modifications in order to get the tests to run. There seems to be code paths that are not covered by the unit tests that might have been broken. I'll try to point the suspects out in the diff.

File: vertica_python/vertica/connection.py [hunk 1]
```diff
@@ -151,10 +151,10 @@ def write(self, message):
 
         except Exception as e:
             self.close_socket()
-            if e.message == 'unsupported authentication method: 9':
+            if str(e) == 'unsupported authentication method: 9':
```
Comment (FredrikAppelros) [inline vertica_python/vertica/connection.py]: I'm not sure this holds. But we can no longer use `message` as that attribute has been removed from `BaseException`.

File: vertica_python/vertica/cursor.py [hunk 2]
```diff
@@ -3,6 +3,8 @@
 import re
 import logging
 
+from builtins import str
```
Comment (FredrikAppelros) [inline vertica_python/vertica/cursor.py]: As per alternative 1 [here](http://python-future.org/compatible_idioms.html#unicode)

File: vertica_python/vertica/column.py [hunk 3]
```diff
@@ -149,6 +150,7 @@ def __init__(self, col):
     def convert(self, s):
         if s is None:
             return
+        s = str(s, 'utf-8')
```
Comment (FredrikAppelros) [inline vertica_python/vertica/column.py]: This needs an explanation, but right now I just have "It seems to work". We might want to do this in a more elegant way.

File: vertica_python/vertica/messages/frontend_messages/query.py [hunk 4]
```diff
@@ -11,11 +11,8 @@ def __init__(self, query_string):
         self.query_string = query_string
 
     def to_bytes(self):
-        s = self.query_string
-        if isinstance(s, str):
-                s = str(s, 'utf-8')
-        encoded = s.encode('utf-8')
+        encoded = self.query_string.encode('utf-8')
```
Comment (FredrikAppelros) [inline vertica_python/vertica/messages/frontend_messages/query.py]: This essentially reverts 6a044c8718ed9ef34458735438fd7b354e286bb6. I'll add a dedicated test to make sure we don't drop support for unicode queries. I think however that the existing test for unicode named parameters already covers this, but better to explicit.

**Comment by nchammas** [comment]
@FredrikAppelros - If the intention is to support both Python 2 and 3 within the same codebase (which I think is the right thing to aim for) then shouldn't you open this PR against `master` instead of `python3-master`? We won't need a separate branch for Python 3 if Python 2 compatibility is maintained.

FWIW, I am ready to test this PR and provide feedback if the library maintainers are on board with the direction.

**Comment by FredrikAppelros** [comment]
I agree that the ideal solution would be to support both Python 2 and 3, and yes, eventually I would assume this to be merged into the `master` branch. But as the maintainers decided to create a separate `python3-master`, I based my fixes on that.

I leave it to the maintainers to merge the `python3-master` branch into `master` at their convenience.

**Comment by BasilPH** [comment]
First a "big thank you" to Fredrik for taking on python 3 support and especially to the maintainers for creating this library in the first place. I would like to use python 3 in my project and this library is keeping me from switching. Is there any movement concerning official python 3 support?
I see that there is a [python 3 package in the ubuntu repositories](https://launchpad.net/ubuntu/wily/+package/python3-vertica). Is there any advantage to using vertica-python over that package?

**Comment by jbfavre** [comment]
Hello,
I have updated my python 3 compatibility patch for release 0.5.6 Debian package.
You can find it here: https://github.com/jbfavre/python-vertica/blob/master/debian/patches/fix_python3_compatibility.patch

Tests are OK for both python 2.7 & 3.5

I'll prepare a pull request against master. I think it's better to have only one branch

**Comment by alexjikim** [comment]
landed this PR for python 3 support: https://github.com/uber/vertica-python/pull/106
